avalanche/evaluation/collector.py

```diff
@@ -0,0 +1,173 @@
+    same stream.
+
+    Example usage:
+
+    .. code-block:: python
+
+        mc = MetricCollector()
+        for exp in train_stream:
+            res = {"Acc": 0.1} # metric dictionary for the current timestep
+            mc.update(res, stream=test_stream)
+        acc_timeline = mc.get("Accuracy", exp_reduce="sample_mean", stream=test_stream)
+
+    """
+
+    def __init__(self):
+        """Init."""
+        self.metrics_res = {}
+
+        self._stream_len = {}  # stream-name -> stream length
+        self._coeffs = {}  # stream-name -> list_of_exp_length
+
+    def _init_stream(self, stream):
+        # we compute the stream length and number of samples in each experience
+        # only the first time we encounter the experience because
+        # iterating over the stream may be expensive
+        self._stream_len[stream.name] = 0
+        coeffs = []
+        for exp in stream:
+            self._stream_len[stream.name] += 1
+            coeffs.append(len(exp.dataset))
+        coeffs = np.array(coeffs)
+        self._coeffs[stream.name] = coeffs / coeffs.sum()
+
+    def update(self, res, *, stream=None):
+        """Update the metrics.
+
+        :param res: a dictionary of new metrics with <metric_name: value> items.
+        :param stream: optional stream. If a stream is given the full metric
+            name becomes `f'{stream.name}/{metric_name}'`.
+        """
+        for k, v in res.items():
+            # optional safety check on metric shape
+            # metrics are expected to have one value for each experience in the stream
+            if stream is not None:
+                if stream.name not in self._stream_len:
+                    self._init_stream(stream)
+                if len(v) != self._stream_len[stream.name]:
+                    raise ValueError(
+                        f"Length does not correspond to stream. "
+                        f"Found {len(v)}, expected {self._stream_len[stream.name]}"
+                    )
+
+            # update metrics dictionary
+            if stream is not None:
+                k = f"{stream.name}/{k}"
+            if k in self.metrics_res:
+                self.metrics_res[k].append(v)
+            else:
+                self.metrics_res[k] = [v]
+
+    def get(self, name, *, time_reduce=None, exp_reduce=None, stream=None):
+        """Returns a metric value given its name and aggregation method.
+
+        :param name: name of the metric.
+        :param time_reduce: Aggregation over the time dimension. One of {None, 'last', 'mean'}, where:
+            - None (default) does not use any aggregation
+            - 'last' returns the last timestep
+            - 'mean' averages over time
+        :param exp_reduce: Aggregation over the experience dimension. One of {None, 'sample_mean', 'experience_mean'} where:
+            - None (default) does not use any aggregation
+            - `sample_mean` is an average weighted by the number of samples in each experience
```

Comment:
Nice ! Maybe we could add a class means too, or is that too specific ? I think that in general in continual learning this option is useful when experiences are not of the same size, and because we often use classical benchmarks that have the same number of samples per class in the test set we often think the sample mean and the class mean is the same, but I think what we want in reality is the class mean in that case. I don't know if you were thinking of another application case. 

I would put both and add something in the doc to say which one should be used in what situation

Reply:
This is a good point. Maybe I can add an argument for a custom reduce function to support this use case. I prefer not to add reduce methods that depend on classification.